cdap-security/src/main/java/co/cask/cdap/security/tools/GeneratedCertKeyStores.java

```diff
@@ -0,0 +1,155 @@
+
+      X509Certificate cert = getCertificate(distinguishedName, pair, validity, signatureAlgo);
+
+      keyStore = KeyStore.getInstance(sConf.get(Constants.Security.SSL.KEYSTORE_TYPE, SSL_KEYSTORE_TYPE));
+      keyStore.load(null, password.toCharArray());
+      keyStore.setKeyEntry(CERT_ALIAS, pair.getPrivate(), password.toCharArray(),
+                           new java.security.cert.Certificate[]{cert});
+    } catch (Exception e) {
+      throw new RuntimeException("SSL is enabled but a key store file could not be created. A keystore is required " +
+                                   "for SSL to be used.", e);
+    }
+    return keyStore;
+  }
+
+  /**
+   * Generate an X.509 certificate
+   *
+   * @param dn Distinguished name for the owner of the certificate, it will also be the signer of the certificate.
+   * @param pair Key pair used for signing the certificate.
+   * @param days Validity of the certificate.
+   * @param algorithm Name of the signature algorithm used.
+   * @return A X.509 certificate
+   */
+  private static X509Certificate getCertificate(String dn, KeyPair pair, int days, String algorithm) {
+    // Calculate the validity interval of the certificate
+    Date from = new Date();
+    Date to = DateUtils.addDays(from, days);
+    CertificateValidity interval = new CertificateValidity(from, to);
+    // Generate a random number to use as the serial number for the certificate
+    BigInteger sn = new BigInteger(64, new SecureRandom());
+    // Create the name of the owner based on the provided distinguished name
+    try {
+      X500Name owner = new X500Name(dn);
+      // Create an info objects with the provided information, which will be used to create the certificate
+      X509CertInfo info = new X509CertInfo();
+      info.set(X509CertInfo.VALIDITY, interval);
+      info.set(X509CertInfo.SERIAL_NUMBER, new CertificateSerialNumber(sn));
+      // This certificate will be self signed, hence the subject and the issuer are same.
+      info.set(X509CertInfo.SUBJECT, new CertificateSubjectName(owner));
+      info.set(X509CertInfo.ISSUER, new CertificateIssuerName(owner));
+      info.set(X509CertInfo.KEY, new CertificateX509Key(pair.getPublic()));
+      info.set(X509CertInfo.VERSION, new CertificateVersion(CertificateVersion.V3));
+      AlgorithmId algo = new AlgorithmId(AlgorithmId.md5WithRSAEncryption_oid);
+      info.set(X509CertInfo.ALGORITHM_ID, new CertificateAlgorithmId(algo));
+      // Create the certificate and sign it with the private key
+      X509CertImpl cert = new X509CertImpl(info);
+      PrivateKey privateKey = pair.getPrivate();
+      cert.sign(privateKey, algorithm);
+      return cert;
+    } catch (CertificateException | IOException | NoSuchAlgorithmException | SignatureException
+      | InvalidKeyException | NoSuchProviderException e) {
+      throw new RuntimeException("SSL is enabled but an certificate could not be generated. A certificate is required" +
```

Comment:
Since this is a private method, it's better to propagate the exception and let caller to handle it, otherwise the exception will get double wrapped.